src/d_iwad.c

```diff
@@ -345,27 +348,245 @@ static void CheckInstallRootPaths(void)
 
 // Check for Doom downloaded via Steam
 
-static void CheckSteamEdition(void)
+static void AddSteamLibraryDir(char *dir)
 {
-    char *install_path;
     char *subpath;
     size_t i;
 
-    install_path = GetRegistryString(&steam_install_location);
+    for (i=0; i<arrlen(steam_install_subdirs); ++i)
+    {
+        subpath = M_StringJoin(dir, DIR_SEPARATOR_S, steam_install_subdirs[i],
+                               NULL);
 
-    if (install_path == NULL)
+        AddIWADDir(subpath);
+    }
+}
+
+//Read whole of libraryfolders.vdf as string
+//silently fail is something goes wrong reading the file
+
+static char *GetSteamLibText(char *library_path)
+{
+    char *library_file_buffer;
+    size_t library_file_size;
+    size_t count;
+    char *probe;
+    FILE *library_file;
+
+    probe = M_FileCaseExists(library_path);
+
+    if (probe == NULL)
+    {
+        return NULL;
+    }
+
+    library_file = fopen(probe, "rb");
+
+    if (library_file == NULL)
+    {
+        return NULL;
+    }
+
+    library_file_size = M_FileLength(library_file);
+    library_file_buffer = malloc(library_file_size + 1);
+    count = fread(library_file_buffer, 1, library_file_size, library_file);
+    fclose(library_file);
+
+    if (library_file_size != count)
+    {
+        free(library_file_buffer);
+        return NULL;
+    }
+
+    library_file_buffer[library_file_size] = '\0';
+    return library_file_buffer;
+}
+
+//Retrieve tokens from file in Valve Data File (VDF) format
+
+static char *GetNextSteamLibToken(char **lib_file_buffer)
+{
+    char c, *buffer, *token;
+    size_t i, j, buff_size, token_size;
+    boolean success, quote;
+
+    token = NULL;  //In VDF a token is anything surrounded by white space.
+                   //Token may or may not be in double quotes.
+                   //Special characters can be escaped with a back slash.
+
+    success = false;
+    buffer = *lib_file_buffer;
+    buff_size = strlen(buffer);
+
+    for (i=0; i<buff_size; ++i)
+    {
+        c = buffer[i];
+
+        //{ and } are control characters in VDF used to indicate subkeys.
+        //If we treat } as white space and allow { to be treated as a regular
+        //token, then we can ignore the concept of subkeys in a VDF file.
+        //This allows us to treat a VDF file as key value pairs.
+
+        if (isspace(c) || c == '}')
+        {
+            continue;
+        }
+
+        quote = c == '"';
+        token = &buffer[i++];
+        break;
+    }
+
+    for (; i<buff_size; ++i)
```

Comment:
This ends up being complicated because you're trying to process different types of things using the same loop. You'd be better off just treating them separately. For example:

```
c = ... whatever
if (c == '"') {
    return ReadStringToken(.....);
} else if (c == '{') {
    return TOKEN_OPEN_BRACE;
} else if (c == '}') {
    return TOKEN_CLOSE_BRACE;
} else if (isalpha(c)) {
    return ReadKeyword(......);
} else {
    return TOKEN_UNKNOWN;
}